proposals/0000-linear-types.rst

```diff
@@ -0,0 +1,2357 @@
+fail to fully consume their argument. We can think of it as: the
+consumption of the result of the function was never complete, so the
+consumption of the argument need not be either. However, because
+exceptions can be caught, a program can observe a state where a value
+``v`` has been passed to a linear function ``f`` but the call ``f v``
+has exited (with an exception) without consuming ``v``. So while, the
+guarantee provided by linear functions holds for converging
+computations, we must weaken it in case of divergence:
+
+- Attempting to consume exactly once ``f v``, when ``f`` is a linear
+  function, will consume ``v`` exactly once if the consumption of ``f
+  v`` converges, and *at most once* if it diverges.
+
+Where "consuming at most once" is defined by induction, like
+"consuming exactly once", but every sub-consumption is optional.
+
+In the paper, we gave a simplified specification of a linear ``IO``
+monad (called ``IOL``) which ignored the issue of exception for the
+sake of simplicity. Can we, still, write a resource-safe ``RIO`` monad
+with linear types despite the added difficulty of exceptions? Yes, as
+this section will show.
+
+Concretely, how do we ensure that the sockets from the example API are
+always closed, even in presence of exceptions? This boils down to how
+the ``RIO`` monad is implemented. Below is a sketch of one possible
+implementation of ``RIO`` (see `here
+<https://github.com/tweag/linear-base/blob/master/src/System/IO/Resource.hs>`_
+for a detailed implementation).
+
+First, note that since Haskell programs are of type ``IO ()``, we need a
+way to run ``RIO`` in an ``IO`` computation, this is provided by the
+function
+
+::
+
+  runRIO :: RIO (Unrestricted a) -> IO a
+
+In order to achieve resource safety in presence of exception, ``runRIO``
+is tasked with releasing any live resource in case of exception.
+
+To implement this, ``RIO`` keeps a table of release actions, to be used
+in case of exceptions. Each resource implemented in the ``RIO``
+abstraction registers a release action in the release action table
+when they are acquired.
+
+If no exception occurs, then all resources have been released by the
+program. In case an exception occurs, the program jumps to the handler
+installed by ``runRIO``, which releases the leftover resources.
+
+An alternative strategy would be to add terminators on every resources
+acquired in ``RIO``. Release in the non-exceptional case would still
+be performed by the program, and the GC would be responsible for
+releasing resources in case of exception. The release in case of
+exception would be, however, less timely.
+
+Can ``RIO`` have a ``catch``?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+It is possible to catch exceptions inside of ``RIO``. But in order to
+ensure resource safety, the type cannot be linear:
+
+::
+
+  catchL :: Exception e
+         => RIO (Unrestricted a)
+	 -> (e -> RIO (Unrestricted a))
+	 -> RIO (Unrestricted a)
```

Comment:
Unless I'm missing something, `catchL io h = liftIO $ catch (runRIO io) (runRIO . h)`, is that right? So it's not a primitive of `RIO`.

Reply:
Can `RIO` have a `liftIO`? Once we have `liftIO` and `runRIO`, then we can create nested `RIO` computations:

```
runRIO $ do
    {A}
    liftIO . runRIO $ {B}
    {C}
```

So what happens if a resource is allocated in A and then deallocated in B? If `RIO` is implemented in Haskell, then I don't see a way for the inner context to be aware of the outer one. That means the deallocator either can't consistently remove its resource from the cleanup table, or has to keep a reference to that table.

Alternately, if `RIO` is implemented with compiler magic, then we essentially have the same choices as with `catch`: either we have a stack of cleanup tables, in which case we potentially have to walk back through the tables to deallocate, or we don't guarantee that anything gets cleaned up until the end of the outer `RIO` computation.

Reply:
@simonmar I believe this is correct (for this first version). Does it matter for the proposal? Since it doesn't actually mention a `liftIO`.

@davemenendez It can. _But_ you have to be careful, `liftIO` need to have type `liftIO :: IO a -> RIO a` (the important part is the unrestricted arrow), so that the `IO` bit cannot capture any of the resources allocated in the enclosing `runRIO` (in your example, you simply can't allocate a resource in A, then deallocated in B. You could deallocate it in C, however). Would you like a paragraph on this point in the proposal?

Reply:
@aspiwack I had been assuming that `RIO` had `liftIO`, since without it you can't do very much. But since one has to be careful in defining `liftIO` as you mentioned, I think it would be good to add it to the proposal.

Now, if you have `liftIO`, you can define `catchL` in terms of it and `runRIO`, it would help to say that `catchL` isn't adding any new functionality (at least the first version of it).